File: server/src/services/user/user.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { User } from 'src/models/user.entity';
import { PasswordUtility } from '../../utility/password/password.service';

@Injectable()
export class UserService {
  private passwordUtility = new PasswordUtility();

  constructor(
    @InjectRepository(User)
    private userRepository: Repository<User>,
  ) { }

  async getUsers(): Promise<User[]> {
    return this.userRepository.find();
  }

  async addUser(userData: { username: string; name: string; password: string }): Promise<User> {
    const user = this.userRepository.create(userData);
    return this.userRepository.save(user);
  }

  async updateUser(userId: string, user: Partial<User>): Promise<User | null> {
    await this.userRepository.update(userId, user);
    return this.userRepository.findOneBy({ id: userId });
  }

  async deleteUser(userId: string): Promise<void> {
    await this.userRepository.delete(userId);
  }

  async findByUsername(username: string): Promise<User | null> {
    return this.userRepository.findOne({ where: { username } });
  }

  async findById(id: string): Promise<User | null> {
    return this.userRepository.findOne({ where: { id } });
  }

  async validateUser(username: string, password: string): Promise<User | null> {
    const user = await this.userRepository.findOne({ where: { username } });

    if (user && user.password === password) {
      return user;
    }

    return null;
  }

}

File: server-java/moneyjar/moneyjar/src/main/java/com/tarungattu/moneyjar/services/user/UserService.java
package com.tarungattu.moneyjar.services.user;

import com.tarungattu.moneyjar.dto.PrivateUserDto;
import com.tarungattu.moneyjar.models.User;
import com.tarungattu.moneyjar.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
public class UserService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    public List<User> getUsers() {
        return userRepository.findAll();
    }

    public User addUser(PrivateUserDto userDto) {
        User user = new User();
        user.setName(userDto.getName());
        user.setUsername(userDto.getUsername());
        user.setPassword(passwordEncoder.encode(userDto.getPassword()));
        return userRepository.save(user);
    }

    public User updateUser(UUID userId, PrivateUserDto userDto) {
        return userRepository.findById(userId).map(user -> {
            if (userDto.getName() != null) user.setName(userDto.getName());
            if (userDto.getUsername() != null) user.setUsername(userDto.getUsername());
            if (userDto.getPassword() != null) user.setPassword(passwordEncoder.encode(userDto.getPassword())); // Optional update?
            // NestJS code: await this.userRepository.update(userId, user); 
            // This implies partial update.
            return userRepository.save(user);
        }).orElse(null);
    }

    public void deleteUser(UUID userId) {
        userRepository.deleteById(userId);
    }

    public User findByUsername(String username) {
        return userRepository.findByUsername(username).orElse(null);
    }

    public User findById(UUID id) {
        return userRepository.findById(id).orElse(null);
    }
}
